tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc

```diff
@@ -0,0 +1,573 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/c/experimental/filesystem/filesystem_interface.h"
+
+#include "google/cloud/storage/client.h"
+
+#include <string>
+#include <memory>
+#include <fstream>
+#include <iostream>
+
+#include <stdio.h>
+
+#include "tensorflow/c/tf_status.h"
+
+// Implementation of a filesystem for GCS environments.
+// This filesystem will support `gs://` URI scheme.
+#define TF_ReturnIfError(status) if(TF_GetCode(status) != TF_OK) return
```

Comment:
Since this is a macro, we should type it accordingly:

```suggestion
#define TF_RETURN_IF_ERROR(status) \
  do { \
    if(TF_GetCode(status) != TF_OK) \
      return; \
  } while(0)
```

I also changed the macro implementation to ensure it cannot be accidentally used in if conditions, basically to sanitize it. 

In fact, I'm wondering why not making it return the status (so it is useful). Let's try to mimick https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/errors.h;l=71-75;drc=16b5232a803abcb55a30fa8326edbe279aca2a1e as much as possible.

(Probably this is moot, as we were planning to move this to another PR, but these comments should also apply there)

Reply:
I have made another PR for this. https://github.com/tensorflow/tensorflow/pull/39319. Those changes will be dropped after that PR is merged. Return the status does not thing because all function in the modular_filesystem does not return a status

Reply:
Oh, but in that case that would be a silent failure. I'm not sure that would be ok.

We should at least print something to the log in that case. I guess I'll have to get to review the call-sites for this and then we can talk about what happens.